backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -409,6 +414,82 @@ public FutureExperimentRunDAO(
         .thenCompose(unused -> deleteExperimentRuns(runIds), executor);
   }
 
+  /**
+   * @param newDatasets : new datasets for privilege check
+   * @param errorOut : Throw error while creation (true) otherwise we will keep it silent (false)
+   * @return {@link List} : accessible datasets
+   */
+  private InternalFuture<List<Artifact>> filterAndGetPrivilegedDatasetsOnly(
+      List<Artifact> newDatasets, boolean errorOut) {
+    List<String> newDatasetLinkedIds = new ArrayList<>();
+    for (Artifact dataset : newDatasets) {
+      String datasetVersionId = dataset.getLinkedArtifactId();
+      if (!datasetVersionId.isEmpty()) {
+        newDatasetLinkedIds.add(datasetVersionId);
+      }
+    }
+
+    if (newDatasetLinkedIds.isEmpty()) {
+      return InternalFuture.completedInternalFuture(newDatasets);
+    }
+    return jdbi.withHandle(
+            handle -> {
+              Map<String, Long> datasetVersionDatasetMap = new HashMap<>();
+              handle
+                  .createQuery(
+                      " SELECT commit_hash, repository_id FROM repository_commit WHERE commit_hash IN (<datasetIds>) ")
+                  .bindList("datasetIds", newDatasetLinkedIds)
+                  .map(
+                      (rs, ctx) ->
+                          datasetVersionDatasetMap.put(
+                              rs.getString("commit_hash"), rs.getLong("repository_id")))
+                  .list();
+              return datasetVersionDatasetMap;
+            })
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              List<String> accessibleDatasetVersionIds = new ArrayList<>();
+              List<InternalFuture<Artifact>> internalFutures = new LinkedList<>();
+              for (Artifact dataset : newDatasets) {
+                String datasetVersionId = dataset.getLinkedArtifactId();
+                if (!datasetVersionId.isEmpty()
+                    && !accessibleDatasetVersionIds.contains(datasetVersionId)) {
+                  try {
+                    if (datasetVersionDatasetMap.containsKey(datasetVersionId)
+                        && datasetVersionDatasetMap.get(datasetVersionId) != null) {
+                      Long datasetId = datasetVersionDatasetMap.get(datasetVersionId);
+
+                      internalFutures.add(
+                          checkEntityPermissionBasedOnResourceTypes(
+                                  Collections.singletonList(String.valueOf(datasetId)),
+                                  ModelDBActionEnum.ModelDBServiceActions.READ,
+                                  ModelDBServiceResourceTypes.DATASET)
+                              .thenCompose(
+                                  unused -> {
+                                    accessibleDatasetVersionIds.add(datasetVersionId);
+                                    return InternalFuture.completedInternalFuture(dataset);
+                                  },
+                                  executor));
```

Comment:
This code is buggy because you're doing `accessibleDatasetVersionIds.add(datasetVersionId)` as a side effect inside the future. This means that you're adding a race condition on the system, which is bad.

Reply:
Changed entire code as per suggestions